src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,362 @@
+  exact is_projection.complement h,
+end ⟩
+
+instance : has_compl (subtype (is_projection  : R → Prop)) :=
+⟨λ P, ⟨1-P, P.prop.complement⟩⟩
+
+end is_projection
+
+variables {X : Type*} [normed_group X]
+
+variables {𝕜 : Type*} [normed_field 𝕜] [normed_space 𝕜 X]
+
+/--
+A projection on a normed space `X` is said to be an L-projection if, for all `x` in `X`,
+$$
+∥x∥ = ∥P x∥ + ∥(1-P) x∥.
+$$
+-/
+def is_Lprojection (P : X →L[𝕜] X) : Prop := is_projection P ∧ ∀ (x : X), ∥x∥ = ∥P x∥ + ∥(1-P) x∥
+
+/--
+A projection on a normed space `X` is said to be an M-projection if, for all `x` in `X`,
+$$
+∥x∥ = max(∥P x∥, ∥(1-P) x∥).
+$$
+-/
+def is_Mprojection (P: X →L[𝕜] X) : Prop :=
+  is_projection P ∧ ∀ (x : X), ∥x∥ = (max ∥P x∥  ∥(1-P) x∥)
+
+namespace is_Lprojection
+
+lemma Lcomplement {P: X →L[𝕜] X} (h: is_Lprojection P) :  is_Lprojection (1-P) :=
+⟨is_projection.complement_iff.mp h.1, λ x, by { rw [add_comm, sub_sub_cancel], exact h.2 x }⟩
+
+lemma Lcomplement_iff (P: X →L[𝕜] X) : is_Lprojection P ↔ is_Lprojection (1-P) :=
+⟨Lcomplement, λ h, by { rw [← sub_sub_cancel 1 P], exact Lcomplement h }⟩
+
+lemma PQ_eq_QPQ (P Q : X →L[𝕜] X) (h₁: is_Lprojection P) (h₂: is_Lprojection Q) :
+  P * Q = Q * P * Q :=
+begin
+  ext,
+  rw ← norm_sub_eq_zero_iff,
+  have e1 : ∥Q x∥ ≥ ∥Q x∥ + 2 • ∥ (P * Q) x - (Q * P * Q) x∥ :=
+  calc ∥Q x∥ = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥(Q * Q) x - Q (P (Q x))∥
+    + ∥(1-Q) ((1 - P) (Q x))∥) :
+    by rw [h₁.right, h₂.right, h₂.right ((1 - P) (Q x)), continuous_linear_map.sub_apply 1 P,
+      continuous_linear_map.one_apply, map_sub, continuous_linear_map.coe_mul]
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥
+    + ∥((1-Q) * Q) x - (1-Q) (P (Q x))∥) : by rw [← sq, projection_def h₂.left,
+      continuous_linear_map.sub_apply 1 P, continuous_linear_map.one_apply,
+      map_sub,continuous_linear_map.coe_mul]
+  ... = ∥Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥ + (∥Q x - Q (P (Q x))∥ + ∥(1-Q) (P (Q x))∥) :
+    by rw [sub_mul, ← sq, projection_def h₂.left, one_mul, sub_self,
+      continuous_linear_map.zero_apply, zero_sub, norm_neg]
+  ... = ∥Q (P (Q x))∥ + ∥Q x - Q (P (Q x))∥ + 2•∥(1-Q) (P (Q x))∥  : by abel
+  ... ≥ ∥Q x∥ + 2 • ∥ (P * Q) x - (Q * P * Q) x∥ :
+    by exact add_le_add_right (norm_le_insert' (Q x) (Q (P (Q x)))) (2•∥(1-Q) (P (Q x))∥),
+  rw ge at e1,
+  nth_rewrite_rhs 0 ← add_zero (∥Q x∥) at e1,
+  rw [add_le_add_iff_left, two_smul,  ← two_mul]  at e1,
+  rw le_antisymm_iff,
+  refine ⟨_, norm_nonneg _⟩,
+  rwa [←mul_zero (2:ℝ), mul_le_mul_left (show (0:ℝ) < 2, by norm_num)] at e1
+end
+
+lemma QP_eq_QPQ (P Q : X →L[𝕜] X) (h₁: is_Lprojection P) (h₂: is_Lprojection Q) : Q * P = Q * P * Q
+  :=
+begin
+  have e1: P * (1 - Q) = P * (1 - Q) - (Q * P - Q * P * Q) :=
+  calc P * (1 - Q) = (1 - Q) * P * (1 - Q) : by rw PQ_eq_QPQ P (1 - Q) h₁ h₂.Lcomplement
+  ... = P * (1 - Q) - (Q * P - Q * P * Q) :
+    by rw [mul_assoc, sub_mul, one_mul, mul_sub, mul_one, mul_sub Q, mul_assoc],
```

Comment:
This is minor, but `noncomm_ring` will solve this for you (since `X →L[𝕜] X` is a non-commutative ring), and perhaps some other places too. You need `import tactic.noncomm_ring`.
```suggestion
  ... = P * (1 - Q) - (Q * P - Q * P * Q) : by noncomm_ring,
```

Reply:
Used this in two places.